daemons/dss-sync-sfn/app.py

```diff
@@ -0,0 +1,306 @@
+"""
+See Readme.md in this directory for documentation on the dss-sync daemon.
+"""
+
+import os, sys, json, logging
+from urllib.parse import unquote
+from concurrent.futures import ThreadPoolExecutor
+
+import domovoi
+
+from dcplib.s3_multipart import AWS_MIN_CHUNK_SIZE
+
+pkg_root = os.path.abspath(os.path.join(os.path.dirname(__file__), 'domovoilib'))  # noqa
+sys.path.insert(0, pkg_root)  # noqa
+
+import dss
+from dss import Config, Replica
+from dss.util.aws import resources, clients
+from dss.logging import configure_lambda_logging
+from dss.events.handlers.sync import (compose_upload, initiate_multipart_upload, complete_multipart_upload, copy_part,
+                                      exists, get_part_size, get_sync_work_state, parts_per_worker, dependencies_exist,
+                                      do_oneshot_copy, sync_sfn_dep_wait_sleep_seconds, sync_sfn_num_threads)
+
+configure_lambda_logging()
+logger = logging.getLogger(__name__)
+dss.Config.set_config(dss.BucketConfig.NORMAL)
+
+app = domovoi.Domovoi()
+
+# This entry point is for S3 native events forwarded through SQS.
+@app.s3_event_handler(bucket=Config.get_s3_bucket(), events=["s3:ObjectCreated:*"], use_sqs=True)
+def launch_from_s3_event(event, context):
+    source_replica = Replica.aws
+    executions = {}
+    if event.get("Event") == "s3:TestEvent":
+        logger.info("S3 test event received and processed successfully")
+    else:
+        for event_record in event["Records"]:
+            bucket = resources.s3.Bucket(event_record["s3"]["bucket"]["name"])
+            obj = bucket.Object(unquote(event_record["s3"]["object"]["key"]))
+            if bucket.name != source_replica.bucket:
+                logger.error("Received S3 event for bucket %s with no configured replica", bucket.name)
+                continue
+
+            for dest_replica in Config.get_replication_destinations(source_replica):
+                if exists(dest_replica, obj.key):
+                    logger.error("Key %s already exists in %s, skipping sync", obj.key, dest_replica)
+                    continue
+                exec_name = bucket.name + "/" + obj.key + ":" + source_replica.name + ":" + dest_replica.name
```

Comment:
It may be useful to include the object type (`blob`, `file`, `bundle`, `collection`) in the `sfn` execution name.

Reply:
Will fix, thanks.

Reply:
The key name starts with `blobs/`, `files/`, `bundles/` or `collections/` already, so the execution name already includes this information.

The execution name only gets printed in the trampoline lambda log output though, it doesn't get passed into the state functions API as the execution name. To pass it in, I'll need to make a change to Domovoi. Filed that for myself as a task. I'd prefer to do that separately from this PR though.

Reply:
Created https://github.com/kislyuk/domovoi/issues/16